keps/sig-network/2079-network-policy-port-range/README.md

```diff
@@ -0,0 +1,339 @@
+
+## Motivation
+
+NetworkPolicy object is a complex object, that allows a developer to specify 
+what's the traffic behavior expected of the application and allow/deny 
+undesired traffic.
+
+There are a number of user issues like [kubernetes #67526](https://github.com/kubernetes/kubernetes/issues/67526) 
+and [kubernetes #93111](https://github.com/kubernetes/kubernetes/issues/93111) 
+where users expose the need to create a policy that allow a range of ports but some 
+specific port, or also cases that a user wants to create a policy that allows 
+the egress to other cluster to the NodePort range (eg 32000-32768) and in this case, 
+the rule should be created specifying each port separately, as:
+
+```
+spec:
+  egress:
+  - ports:
+    - protocol: TCP
+      port: 32000
+    - protocol: TCP
+      port: 32001
+    - protocol: TCP
+      port: 32002
+    - protocol: TCP
+      port: 32003
+[...]
+    - protocol: TCP
+      port: 32768
+```
+
+So for the user:
+* To allow a range of ports, each of them must be declared as an item from 
+``ports`` array
+* To make an exception needs a declaration of all ports but the exception
+
+Adding a new ``PortRange`` field inside the ``ports`` will allow a simpler 
+creation of NetworkPolicy to the user.
+
+### Goals
+
+Add Range field to Ports in NetworkPolicy
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+In NetworkPolicy specification, inside ``NetworkPolicyPort`` object struct 
+specify a new ``Range`` field composed of the minimum and maximum ports 
+inside the range.
+
+If both ``Port`` and ``Range`` are specified, they are cumulative and no further
+validation might occur. It's up to the CNI to summarize both of the fields in a 
+single rule.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I have an application that communicates with NodePorts of a different cluster 
+and I want to allow the egress of the traffic only the NodePort range 
+(eg. 30000-32767) as I don't know which port is going to be allocated on the 
+other side, but don't want to create a rule for each of them.
+
+### Notes/Constraints/Caveats
+
+*  The technology used by the CNI provider might not support port range in a 
+trivial way as described in [#drawbacks]
+
+### Risks and Mitigations
+
+CNIs will need to support the new field in their controllers. For this case 
+we'll try to make broader communication with the main CNIs so they can be aware
+of the new field.
+
+## Design Details
+
+API changes to NetworkPolicy:
+* Add a new struct called ``NetworkPolicyPortRange`` as the following:
+```
+// NetworkPolicyPortRange describes the range of ports to be used in a 
+// NetworkPolicyPort struct
+type NetworkPolicyPortRange struct {
+    // From defines the start of the port range
+    From         uint16
+    
+    // To defines the end of the port range, being the end included within the
+    // range
+    To           uint16
+
+    // Except defines all the exceptions in the port range
+    Except:      []uint16
```

Comment:
wait, I thought exceptions were going away?

Reply:
@danwinship Last policy api meeting we've decided that the efforts to add exceptions again is too low compared to the benefits of adding it in the struct.

Basically, it's clearer when you define an exception inside a range instead of defining multiple ranges excluding an exception, like:

```
- from: 32000
  to: 32999
- from: 33001 # Excluding the 33000
  to: 34000
```
vs
```
- from: 32000
  to: 34000
  except:
  - 33001
```

So I've reinserted this in the struct, and removed this as a non goal :)

Reply:
Well, ok, but what is the use case for exceptions then, such that they need to be easy to express? What is the developer use case for granting a pod access to 32000-34000 _except_ 33001?

There were strong objections to adding `except` to `ipBlock`, but we kept it because it's incredibly difficult to do the equivalent of

    - from: 32000
      to: 32999
    - from: 33001
      to: 34000

with CIDRs. if we didn't have `except` then in order to write

    - ipBlock:
        cidr: 192.168.0.0/24
        except:
          - 192.168.0.5

you'd have to do:

    - cidr: 192.168.0.128/25
    - cidr: 192.168.0.64/26
    - cidr: 192.168.0.32/27
    - cidr: 192.168.0.16/28
    - cidr: 192.168.0.8/29
    - cidr: 192.168.0.6/31
    - cidr: 192.168.0.4/32
    - cidr: 192.168.0.2/31
    - cidr: 192.168.0.1/32

which is both horrifying and incomprehensible (and probably also wrong; I didn't double-check it).

Reply:
I think this user story is still compliant, even if we've removed the exception, thinking that the user can create multiple ranges excluding the excepts, right?